M4A/AAC: Fix seeking backwards

uklotzde opened the pull request:
While working on the SoundSourceMediaFoundation plugin I noticed that we don't have a test for seeking backwards. I extended the existing test and it turned out that FAAD2 is not sample accurate when seeking to a position near the beginning of the stream.

Workaround/Fix: Reopen the AAC decoder when seeking to a position near the beginning of the stream.

Update: Discovered and fixed a bug in SoundSourceMediaFoundation when seeking backwards to the beginning of the stream.

uklotzde committed:
Regression test: Seek both forwards and backwards

uklotzde committed:
SoundSourceM4A: Reorder member variables

uklotzde committed:
SoundSourceM4A: Open AAC decoder in separate member function

uklotzde committed:
SoundSourceM4A: Add workaround for seeking backwards

daschuer commented:
Nice catch. LGTM! Thank you.

daschuer commented:
```
plugins/soundsourcem4a/soundsourcem4a.cpp:242:19: error: use of deleted function ‘mixxx::AudioSourceConfig& mixxx::AudioSourceConfig::operator=(const mixxx::AudioSourceConfig&)’
     m_audioSrcCfg = audioSrcCfg;
                   ^
In file included from src/sources/soundsource.h:5:0,
                 from src/sources/soundsourceprovider.h:8,
                 from src/sources/soundsourcepluginapi.h:30,
                 from src/sources/soundsourceplugin.h:4,
                 from plugins/soundsourcem4a/soundsourcem4a.h:4,
                 from plugins/soundsourcem4a/soundsourcem4a.cpp:1:
src/sources/audiosource.h:218:7: note: ‘mixxx::AudioSourceConfig& mixxx::AudioSourceConfig::operator=(const mixxx::AudioSourceConfig&)’ is implicitly deleted because the default definition would be ill-formed:
```

uklotzde committed:
Revert explicit declaration of move/copy ctors for AudioSignal

As expected we should not declare move constructors and assignment
operators until we need them. Copy construction and assignment is
all that we need here.

uklotzde commented:
Bingo! One more test and everything falls apart ;) I will fix the Windows failures, but I don't know what went wrong for OSX.

rryan commented:
> but I don't know what went wrong for OSX.

It also affects master (I think it was caused by #887 which I just merged from 1.12  https://github.com/mixxxdj/mixxx/commit/f3905bab698a9cedadab96ed85cb36697e6035ff), but I can't reproduce locally on OSX. I'm getting debug info in #1121.

uklotzde committed:
Use AudioSource::getMinFrameIndex() instead of 0

...actually it is 0 and the compiler will optimize this for us. Just to
remove any implicit assumptions.

uklotzde committed:
SoundSourceMediaFoundation: Fix seeking to beginning of stream

uklotzde commented:
I've found some more bugs after adding a test for skipping sample frames. Those fixes will follow after this PR has been merged.
